ipinfo/cmd_download.go

```diff
@@ -0,0 +1,227 @@
+    # Download country database in csv format.
+    $ %[1]s download country -f csv > country.csv
+    $ %[1]s download country-asn country_asn.mmdb
+
+Databases:
+    asn            free ipinfo asn database.
+    country        free ipinfo country database.
+    country-asn    free ipinfo country-asn database.
+
+Options:
+  General:
+    --token <tok>, -t <tok>
+      use <tok> as API token.
+    --help, -h
+      show help.
+
+Outputs:
+    --compress, -c
+	save the file in compressed format.
+	default: false.
+    --format, -f <mmdb | json | csv>
+     output format of the database file.
+     default: mmdb.
+`, progBase)
+}
+
+func cmdDownload() error {
+	var fTok string
+	var fFmt string
+	var fZip bool
+	var fHelp bool
+
+	pflag.StringVarP(&fTok, "token", "t", "", "the token to use.")
+	pflag.StringVarP(&fFmt, "format", "f", "mmdb", "the output format to use.")
+	pflag.BoolVarP(&fZip, "compress", "c", false, "compressed output.")
+	pflag.BoolVarP(&fHelp, "help", "h", false, "show help.")
+	pflag.Parse()
+
+	args := pflag.Args()[1:]
+	if fHelp || len(args) > 2 || len(args) < 1 {
+		printHelpDownload()
+		return nil
+	}
+
+	var token string
+	if fTok == "" {
+		token = gConfig.Token
+	}
+
+	// require token for download.
+	if token == "" {
+		return errors.New("downloading requires a token; login via `ipinfo login` or pass the `--token` argument")
+	}
+
+	// get download format and extension.
+	var format string
+	var fileExtension string
+	switch strings.ToLower(fFmt) {
+	case "mmdb":
+		format = "mmdb"
+		fileExtension = "mmdb"
+	case "csv":
+		format = "csv.gz"
+		fileExtension = "csv"
+	case "json":
+		format = "json.gz"
+		fileExtension = "json"
+	default:
+		return errors.New("unknown download format")
+	}
+
+	if fZip {
+		fileExtension = fmt.Sprintf("%s.%s", fileExtension, "gz")
+	}
+
+	// download the db.
+	switch strings.ToLower(args[0]) {
+	case "asn":
+		err := downloadDb("asn", format, token, fileExtension, fZip)
+		if err != nil {
+			return err
+		}
+	case "country":
+		err := downloadDb("country", format, token, fileExtension, fZip)
+		if err != nil {
+			return err
+		}
+	case "country-asn":
+		err := downloadDb("country_asn", format, token, fileExtension, fZip)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("database '%v' is invalid", args[0])
+	}
+
+	return nil
+}
+
+func downloadDb(name, format, token, fileExtension string, zip bool) error {
+	url := fmt.Sprintf("%s%s.%s?token=%s", dbDownloadURL, name, format, token)
+
+	// get file name.
+	var fileName string
+	if len(pflag.Args()) > 2 {
+		fileName = pflag.Args()[2]
+	} else {
+		fileName = fmt.Sprintf("%s.%s", name, fileExtension)
+	}
+
+	// make API req to download the file.
+	res, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	// if output not terminal unzip and write to stdout.
+	if fileInfo, _ := os.Stdout.Stat(); (fileInfo.Mode() & os.ModeCharDevice) == 0 {
+		err := unzipWrite(os.Stdout, res.Body)
```

Comment:
What about if the user passes `-c` for this case?

Reply:
I thought that'll be overkill as I couldn't imagine any realistic scenario where one wants to pipe stdout to some zip program. The maximum someone will do is redirect the output to a file maybe `>` and for that we are already using the argument for the file name.

Reply:
I can imagine a very important use case I use all the time - `ipinfo download country-asn -c | zcat | head -n10` - this shows me the first 10 entries in country-asn, and allows me to download the zipped version of those.